px.py

```diff
@@ -347,24 +365,22 @@ def log_message(self, format, *args):
         dprint(format % args)
 
     def do_socket_connect(self, destination=None):
+        # Already connected?
         if hasattr(self, "proxy_socket") and self.proxy_socket is not None:
             return True
 
-        dest = State.proxy_server
-        if destination is not None:
-            dest = [destination]
-
-        for i in range(len(dest)):
-            dprint("New connection: " + str(dest[0]))
-            self.proxy_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.proxy_socket = None
+        dests = self.proxy_servers if destination is None else [destination]
+        for dest in dests:
+            dprint("New connection: " + str(dest))
+            proxy_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
             try:
-                self.proxy_socket.connect(dest[0])
-                self.proxy_address = dest[0]
+                proxy_socket.connect(dest)
+                self.proxy_address = dest
+                self.proxy_socket = proxy_socket
                 break
-            except:
-                dprint("Connect failed")
-                dest.append(dest.pop(0))
```

Comment:
This is in place because you want to make the proxy server that we connected to successfully in the last attempt as the first in the list so that you don't keep trying to connect to failing proxies earlier in the list after we know a proxy didn't work. That's only possible if you loop over the indexes rather than the array directly and always refer to dest[0]. New code doesn't do this so if your primary proxy fails, Px will keep trying to connect to it and waste time.

Reply:
Ok, intention is clear now. Unfortunately index method could lead to wrong behaviour with manual configured proxies due to threading. For autoconf each thread got its own self.proxy_servers from find_proxy_for url(). No problem. But for manual configuration the proxy list is shared between threads (i.e. self.proxy_servers references the same list as the shared State.proxy_server). In the worst case with two proxies in the list two threads could try the first proxy at the same time and fail. They both remove what they see as the first entry from the list (i.e. together they clear the list) and just before they can append the entries again the third thread kicks in and got an empty list. To avoid this we could use

```python
dests = list(self.proxy_servers) if ...
for dest in dests:
...
    except:
        if len(self.proxy_servers) > 1:
            self.proxy_servers.append(dest)
            self.proxy_servers.remove(dest)
```
Here in the worst case a thread could copy and then try a list where a failed proxy was appended multiple times by other threads before they can remove the first occurence. But a thread never gets an empty list and it didn't get to the failed proxies at the end of the list if one of the others in the list is working.
Another possibility is locking. But this would slow down even with no race condition and I would rather not use it here and prefer the solution above. Would it be ok for you?

Reply:
Yes this works.